vscode-extensions.charliermarsh.ruff: replace the bundled ruff binary

B4dM4n opened the pull request:
The extension ships with a precompiled binary by default, which is not compatible with NixOS.


## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

drupol commented:
Hmm, I don't know what to think here. This only happens when Ruff is not available from the shell where VSCode was run. When Ruff is available, the bundled version is not even used.

B4dM4n commented:
It ships with a bundled binary, because there it's not guaranteed that the environment will provide one and so it can fallback to a default version.

Maybe setting the config default was not the right choice. What if we just replace the bundled binary with a symlink, so no settings are modified and the behavior is the same as before, but with a working binary:

```nix
  postInstall = ''
    # check that the binary was not moved away before overriding
    test -x "$out/$installPrefix/bundled/libs/bin/ruff"
    ln -sf ${lib.getExe ruff} "$out/$installPrefix/bundled/libs/bin/ruff"
  '';
```

drupol commented:
What if the shipped version does not correspond to the ruff version used in the python project ?

Azd325 commented:
> Is it using the shipped version or the version found in the path ?

https://docs.astral.sh/ruff/editors/settings/#importstrategy

`fromEnvironment` finds Ruff in the environment, falling back to the bundled version

B4dM4n commented:
The behavior can be configured. The logic is defined here: https://github.com/astral-sh/ruff-vscode/blob/9da8dc269fa836653d51a8d8ad73f823c3238d20/src/common/server.ts#L86-L100

> 1. If the 'path' setting is set, check each path in order. The first valid
>    path is returned.
> 2. If the 'importStrategy' setting is 'useBundled', return the bundled
>    executable path.
> 3. Execute a Python script that tries to locate the binary. This uses either
>    the user-provided interpreter or the interpreter provided by the Python
>    extension.
> 4. If the Python script doesn't return a path, check the global environment
>    which checks the PATH environment variable.
> 5. If all else fails, return the bundled executable path.

So by default (`importStrategy=fromEnvironment`) it would prefer the environment version over the bundled one.

There is an additional check before the first one, if the workspace is untrusted. In this case it will always use the bundled version or fail otherwise.

So replacing the broken binary with a working symlink should be the correct solution, to match the upstream experience.

drupol commented:
Oh then it's super fine ! Thanks for doing the investigation.

B4dM4n committed:
vscode-extensions.charliermarsh.ruff: replace the bundled ruff binary

The extension ships with a precompiled binary by default, which is not compatible with NixOS.

Azd325 commented:
Thanks for finding it.
I can't test it right now, but what is the issue with the bundled version? Could an update of the package fix it? The bundled version is already some days old. 

Do we need to symlink `ruff-lsp` too?
